### DISCUSSION TITLE
Any plan to take advantage of System.Memory

### DISCUSSION THREAD

**Discussion by hanabi1224** [mathnet/mathnet-numerics#828] [Ideas]
Read about that new System.Span and System.Memory brings huge performance gain in .net core, just curious if those would benefit the array operations here?

**Comment by Apollo3zehn** [comment]
Support for Span<T> would be really great and would save me a huge number of copy operations. Implementation should'nt be too difficult. In my case I am interested in functions like
 
- ArrayStatistics.Mean
- ArrayStatistics.Minimum
- ArrayStatistics.Maximum
- ArrayStatistics.StandardDeviation
- ArrayStatistics.RootMeanSquare

I am downsampling data by dividing the original dataset into pieces that are then processed by one of these methods. But for each piece I need to create a new array, with ```Span<T>``` this won't be required anymore.

**Comment by cdrnet** [comment]
Yes, this is on the roadmap!

**Comment by aaronhudon** [comment]
Suggesting `MathNet.Numerics.Fit` functions be considered for `Span` support.

**Comment by ghost** [comment]
I am also interested in this topic. @cdrnet, could you please elaborate on the roadmap you mentioned?

**Comment by RossMMason** [comment]
Any update on this?

**Comment by matthewkolbe** [comment]
I've been hacking away experimentally on how to improve many of array allocation issues that 5.0 and greater solve, but the fixes run deep, so in order not to destroy compatibility with net48 compilation, there would many large swaths of mostly duplicated code guarded by ```#if NET5_0_OR_GREATER```. For example, I'm looking at ```CopyToUnchecked``` int ```DenseVectorStorage```, and it would end up looking something like this:

```           
            if (target is SparseVectorStorage<T> sparseTarget)
            {
#if NET5_0_OR_GREATER
                Span<int> indices = stackalloc int[Data.Length];
                Span<T> values = GC.AllocateUninitializedArray<T>(Data.Length);
#else
                var indices = new int[Data.Length];
                var values = new T[Data.Length];
#endif

                int c = 0;

                for (int i = 0; i < Data.Length; i++)
                {
                    var item = Data[i];
                    if (!Zero.Equals(item))
                    {
                        values[c] = item;
                        indices[c] = i;
                        c++;
                    }
                }
#if NET5_0_OR_GREATER
                sparseTarget.Indices = indices.Slice(0, c).ToArray();
                sparseTarget.Values = values.Slice(0, c).ToArray();
#else
                sparseTarget.Indices = indices.Take(c).ToArray();
                sparseTarget.Values = values.Take(c).ToArray();
#endif
                sparseTarget.ValueCount = c;
                return;
            }
```

@cdrnet, does this seem like the direction you'd want to go? Or just leave these things be?

**Reply by nodakai** [reply]
Forgive my ignorance as I only use the latest .NET versions on Linux servers, but wouldn't https://www.nuget.org/packages/System.Memory/ resolve compat issues for all .NET Frameworks that haven't reached EoL yet?

https://dotnet.microsoft.com/en-us/platform/support/policy/dotnet-framework
https://github.com/dotnet/corefxlab/issues/2581
